pkg/dockerregistry/server/digestcache.go

```diff
@@ -52,42 +64,122 @@ func (c digestToRepositoryCache) RepositoriesForDigest(dgst digest.Digest) []str
 	return repos.Copy()
 }
 
+// repositoryBucket contains a list of repositories with eviction timeouts.
 type repositoryBucket struct {
 	mu   sync.Mutex
-	list []string
+	list []bucketEntry
 }
 
 // Has returns true if the bucket contains this repository.
-func (i *repositoryBucket) Has(repo string) bool {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	for _, s := range i.list {
-		if s == repo {
-			return true
+func (b *repositoryBucket) Has(repo string) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+
+	return b.getIndexOf(repo, &now) >= 0
+}
+
+// Add one or more repositories to this bucket.
+func (b *repositoryBucket) Add(ttl time.Duration, repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+	nowPtr := &now
+	evictOn := now.Add(ttl)
+
+	for _, repo := range repos {
+		index := b.getIndexOf(repo, nowPtr)
+		nowPtr = nil // evict just on the first run
+		arr := b.list
+
+		if index >= 0 {
+			// repository already exists, move it to the end with highest eviction time
+			entry := arr[index]
+			copy(arr[index:], arr[index+1:])
+			if entry.evictOn.Before(evictOn) {
+				entry.evictOn = evictOn
+			}
+			arr[len(arr)-1] = entry
+
+		} else {
+			// repository is a new entry
+			if len(arr) >= bucketSize {
+				arr = arr[1:]
+			}
+			arr = append(arr, bucketEntry{
+				repository: repo,
+				evictOn:    evictOn,
+			})
 		}
+
+		b.list = arr
 	}
-	return false
 }
 
-// Add one or more repositories to this bucket.
-func (i *repositoryBucket) Add(repos ...string) {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	arr := i.list
+// Remove removes all the given repos from repository bucket.
+func (b *repositoryBucket) Remove(repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+	nowPtr := &now
+
 	for _, repo := range repos {
-		if len(arr) >= bucketSize {
-			arr = arr[1:]
+		index := b.getIndexOf(repo, nowPtr)
+		nowPtr = nil // evict just on the first run
+
+		if index >= 0 {
+			copy(b.list[index:], b.list[index+1:])
+			b.list = b.list[:len(b.list)-1]
 		}
-		arr = append(arr, repo)
 	}
-	i.list = arr
+}
+
+// getIndexOf returns an index of given repository in bucket's array. If not found, -1 will be returned.
+// If now is given, all the entries with eviction time before now will be removed.
+func (b *repositoryBucket) getIndexOf(repo string, now *time.Time) int {
```

Comment:
getIndexOf mutating the list is very unexpected. split this into two functions, one that searches for a repo, and one that evicts things with expired TTLs